### PR TITLE
optee imx8 support

### PR DESCRIPTION
**Pull request by petegriffin**
This pull request adds the optee iMX forks. This has been done using the PREFERRED_VERSION method for enabling the fork, similar to what is done already for Weston and Wayland imx forks. This means it can co-exist nicely in builds that include multiple bsp layers that also have optee rather than having a different package name.

Tested on iMX8M-evk board. op-tee Xtest test suite reports

+-----------------------------------------------------
16103 subtests of which 3 failed
75 test cases of which 1 failed
0 test case was skipped
TEE test application done!

### CODE REVIEW

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+# as long as the compilers are correctly defined.
+# For 64bits, CROSS_COMPILE64 must be set
+# When defining CROSS_COMPILE and CROSS_COMPILE64, we assure that
+# any 32 or 64 bits builds will pass
+EXTRA_OEMAKE = "PLATFORM=imx PLATFORM_FLAVOR=${PLATFORM_FLAVOR} \
+                CROSS_COMPILE=${HOST_PREFIX} \
+                CROSS_COMPILE64=${HOST_PREFIX} \
+                NOWERROR=1 \
+                LDFLAGS= \
+		O=${B} \
```
Comment (otavio) [inline recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: Misalignment.

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    for f in ${B}/export-ta_${OPTEE_ARCH}/*; do
+        cp -aR $f ${D}/usr/include/optee/export-user_ta_${OPTEE_ARCH}/
+    done
+}
+
+addtask deploy after do_compile before do_install
+
+
+FILES_${PN} = "/lib/firmware/"
+FILES_${PN}-dev = "/usr/include/optee"
```
Comment (otavio) [inline recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: Why default does not work?

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+INSANE_SKIP_${PN}-dev = "staticdev"
```
Comment (otavio) [inline recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: Why?

File: conf/machine/imx8mqevk.conf [hunk 4]
```diff
@@ -9,7 +9,7 @@ MACHINEOVERRIDES =. "mx8:mx8m:mx8mq:"
 require conf/machine/include/imx-base.inc
 require conf/machine/include/arm/arch-arm64.inc
 
-MACHINE_FEATURES += "pci wifi bluetooth optee qca6174"
+MACHINE_FEATURES += "pci wifi bluetooth qca6174 optee"
```
Comment (otavio) [inline conf/machine/imx8mqevk.conf]: It seems like an unnecessary change.

File: recipes-security/optee-imx/optee-test/0001-regression-4011-correct-potential-overflow.patch [hunk 5]
```diff
@@ -0,0 +1,72 @@
+Upstream-Status: Backport
```
Comment (otavio) [inline recipes-security/optee-imx/optee-test/0001-regression-4011-correct-potential-overflow.patch]: Please mention `[release]`

File: recipes-security/optee-imx/optee-test/0001-xtest-prevent-unexpected-build-warning-with-strncpy.patch [hunk 6]
```diff
@@ -0,0 +1,66 @@
+Upstream-Status: Backport
```
Comment (otavio) [inline recipes-security/optee-imx/optee-test/0001-xtest-prevent-unexpected-build-warning-with-strncpy.patch]: Likewise.

File: recipes-security/optee-imx/optee-test_3.2.0.imx.bb [hunk 7]
```diff
@@ -0,0 +1,57 @@
+# Copyright (C) 2017-2018 NXP
+
+SUMMARY = "OPTEE test"
+HOMEPAGE = "http://www.optee.org/"
+
+LICENSE = "BSD"
+LIC_FILES_CHKSUM = "file://LICENSE.md;md5=daa2bcccc666345ab8940aab1315a4fa"
+
```
Comment (otavio) [inline recipes-security/optee-imx/optee-test_3.2.0.imx.bb]: `DEPENDS` comes before `inherit`.

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb (hunk 1 above)
Comment (petegriffin) [reply recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: will fix

File: conf/machine/imx8mqevk.conf (hunk 4 above)
Comment (petegriffin) [reply conf/machine/imx8mqevk.conf]: will fix

File: recipes-security/optee-imx/optee-test/0001-regression-4011-correct-potential-overflow.patch (hunk 5 above)
Comment (petegriffin) [reply recipes-security/optee-imx/optee-test/0001-regression-4011-correct-potential-overflow.patch]: will fix

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb (hunk 2 above)
Comment (petegriffin) [reply recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: It looks like this is copied from the original meta-linaro optee-os recipe which also includes this (since the first committed git revision from what I can see). Without this I get the following QA warning

WARNING: optee-os-3.2.0.imx-r0 do_package: QA Issue: optee-os: Files/directories were installed but not shipped in any package:
  /lib
  /lib/firmware
  /lib/firmware/tee-pager_v2.bin
  /lib/firmware/tee-pager.bin
  /lib/firmware/tee.bin
  /lib/firmware/tee-pageable.bin
  /lib/firmware/tee-pageable_v2.bin
  /lib/firmware/tee-header_v2.bin
Please set FILES such that these items are packaged. Alternatively if they are unneeded, avoid installing them or delete them within do_install.
optee-os: 8 installed and not shipped files. [installed-vs-shipped]

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb (hunk 3 above)
Comment (petegriffin) [reply recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: Without disabling staticdev I get the following QA warning. Again it appears copied from the Linaro optee recipe and has been present since the first git version [1].

[1] https://git.linaro.org/openembedded/meta-linaro.git/commit/meta-optee/recipes-security/optee/optee-os_git.bb?id=a5d6db26e1bf6a398fcaf87d5a8f2365cbc54090

WARNING: optee-os-3.2.0.imx-r0 do_package_qa: QA Issue: non -staticdev package contains static .a library: optee-os-dev path '/work/imx8mqevk-linaro-linux/optee-os/3.2.0.imx-r0/packages-split/optee-os-dev/usr/include/optee/export-user_ta_arm64/lib/libmbedtls.a'
non -staticdev package contains static .a library: optee-os-dev path '/work/imx8mqevk-linaro-linux/optee-os/3.2.0.imx-r0/packages-split/optee-os-dev/usr/include/optee/export-user_ta_arm64/lib/libutee.a'
non -staticdev package contains static .a library: optee-os-dev path '/work/imx8mqevk-linaro-linux/optee-os/3.2.0.imx-r0/packages-split/optee-os-dev/usr/include/optee/export-user_ta_arm64/lib/libutils.a'
non -staticdev package contains static .a library: optee-os-dev path '/work/imx8mqevk-linaro-linux/optee-os/3.2.0.imx-r0/packages-split/optee-os-dev/usr/include/optee/export-user_ta_arm64/lib/libmpa.a' [staticdev]

File: recipes-security/optee-imx/optee-test/0001-xtest-prevent-unexpected-build-warning-with-strncpy.patch (hunk 6 above)
Comment (petegriffin) [reply recipes-security/optee-imx/optee-test/0001-xtest-prevent-unexpected-build-warning-with-strncpy.patch]: will fix

File: recipes-security/optee-imx/optee-test_3.2.0.imx.bb (hunk 7 above)
Comment (petegriffin) [reply recipes-security/optee-imx/optee-test_3.2.0.imx.bb]: Forgot this one in the second revision, just pushed a third which includes this.

**Comment by petegriffin** [comment]
@otavio I've fixed everything requested in your review apart form
FILES_${PN} = "/lib/firmware/"
FILES_${PN}-dev = "/usr/include/optee"
INSANE_SKIP_${PN}-dev = "staticdev"

Which generate QA warnings if removed.

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb (hunk 2 above)
Comment (otavio) [reply recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: The `FILES_${PN}` seems ok but the `FILES_${PN}-dev` are not required.

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb (hunk 3 above)
Comment (otavio) [reply recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: Those `.a` files ought to be on `FILES_${PN}-staticdev` and a `RDEPENDS_${PN}-dev += "${PN}-staticdev" to be added, so it is broght when doing development when using a SDK.

**Comment by otavio** [comment]
> @otavio I've fixed everything requested in your review apart form
> FILES_${PN} = "/lib/firmware/"
> FILES_${PN}-dev = "/usr/include/optee"
> INSANE_SKIP_${PN}-dev = "staticdev"
> 
> Which generate QA warnings if removed.

I made comments explaining how it could be fixed. Please take a look.

File: recipes-security/optee-imx/optee-os_3.2.0.imx.bb (hunk 2 above)
Comment (petegriffin) [reply recipes-security/optee-imx/optee-os_3.2.0.imx.bb]: ok will remove FILES_${PN}-dev in next revision

**Comment by petegriffin** [comment]
> > @otavio I've fixed everything requested in your review apart form
> > FILES_${PN} = "/lib/firmware/"
> > FILES_${PN}-dev = "/usr/include/optee"
> > INSANE_SKIP_${PN}-dev = "staticdev"
> > Which generate QA warnings if removed.
> 
> I made comments explaining how it could be fixed. Please take a look.

Brilliant, thanks for your advice. I'm just about to push a new version that incorporates that feedback.

**Comment by otavio** [comment]
Please rebase this against current master.

**Comment by petegriffin** [comment]
> Please rebase this against current master.

Done. I just also re-built and reflashed and xtest (op-tee test suite) still passes

+-----------------------------------------------------
16103 subtests of which 3 failed
75 test cases of which 1 failed
0 test case was skipped
TEE test application done!

**Comment by petegriffin** [comment]
@otavio I forgot to say, is it possible to cherry-pick this functionality to Zeus as well?

**Comment by otavio** [comment]
> @otavio I forgot to say, is it possible to cherry-pick this functionality to Zeus as well?

Done.
